File: firebase-api-library/src/main/java/com/fazemeright/library/api/result/Result.java
package com.fazemeright.library.api.result;

import javax.annotation.Nullable;

public class Result<T> implements TaskResult<T> {
  private final T data;
  private final Exception exception;

  private Result(T data) {
    this.data = data;
    exception = null;
  }

  private Result(Exception exception) {
    this.exception = exception;
    data = null;
  }

  public static Result<Void> nullResult() {
    return new Result<>(null);
  }

  public static <R> Result<R> exception(Exception exception) {
    return new Result<R>(exception);
  }

  public static <R> Result<R> withData(R data) {
    return new Result<>(data);
  }

  @Override
  public boolean isSuccessful() {
    return data != null;
  }

  @Override
  public boolean isFailed() {
    return !this.isSuccessful();
  }

  @Nullable
  @Override
  public T getData() {
    return data;
  }

  @Nullable
  @Override
  public Exception getException() {
    return exception;
  }
}

File: firebase-api-library/src/main/java/com/fazemeright/library/api/result/Result.kt
package com.fazemeright.library.api.result

sealed class Result<out T : Any> {

    data class Success<out T : Any>(val data: T, val msg: String = "") : Result<T>()
    data class Error(val exception: Exception? = null, val msg: String? = exception?.message) : Result<Nothing>()

    override fun toString(): String {
        return when (this) {
            is Success<*> -> "Success[data=$data, msg=$msg]"
            is Error -> "Error[exception=$exception, msg=$msg]"
        }
    }
}